hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSource.java

```diff
@@ -583,16 +614,27 @@ private void initialize() {
       PriorityBlockingQueue<Path> queue = entry.getValue();
       tryStartNewShipper(walGroupId, queue);
     }
+    this.startupOngoing.set(false);
   }
 
   @Override
   public void startup() {
     // mark we are running now
     this.sourceRunning = true;
-    initThread = new Thread(this::initialize);
-    Threads.setDaemonThreadRunning(initThread,
-      Thread.currentThread().getName() + ".replicationSource," + this.queueId,
-      this::uncaughtException);
+    this.retryStartup.set(true);
+    do {
+      if(retryStartup.get()) {
+        retryStartup.set(false);
+        startupOngoing.set(true);
+        initThread = new Thread(this::initialize);
+        Threads.setDaemonThreadRunning(initThread,
+          Thread.currentThread().getName() + ".replicationSource," + this.queueId,
+          (t,e) -> {
+          uncaughtException(t, e);
+          retryStartup.set(true);
```

Comment:
>I hope if we encounter uncaughtException here, we want to retry the loop again.

Yes, we want to keep trying until we succeed.

> If so, shall we also add startupOngoing.set(true); here explicitely? Just in case if it is set to false in initialize()?

The only cases `startupOngoing` would had been set to `false` in  `initialize` is if it completes fine without any uncaught exception, so adding it here would be redundant.

Reply:
Yeah right, it is not set to `false` if it completes successfully, anyone touching the same code in future should realize this. 
Although not a strong point but If you don't mind, maybe we can comment here indicating `startupOngoing` is expected to be true when we are here handling Exception.

Reply:
Addressed the nits and added some comments to the two flags.